pkg/agent/runner.go

```diff
@@ -346,17 +354,24 @@ func (r *Runner) spawnBackgroundWorker(ctx context.Context, logger *zap.Logger,
 // getMetricsLoop repeatedly attempts to fetch metrics from the VM
 //
 // Every time metrics are successfully fetched, the value is recorded with newMetrics.
-func (r *Runner) getMetricsLoop(
+func getMetricsLoop[M core.FromPrometheus[C], C any](
+	r *Runner,
 	ctx context.Context,
 	logger *zap.Logger,
-	newMetrics func(metrics core.Metrics, withLock func()),
+	kind string,
+	emptyMetrics func() M,
```

Comment:
Can you create empty M, and drop this arg?

Reply:
Unfortunately `M` is basically always a pointer, which makes this difficult.

We _can_ solve the issue by further convoluting the generics — maybe it's worthwhile, maybe it's not. That would look something like:

```diff
diff --git a/pkg/agent/runner.go b/pkg/agent/runner.go
index 52cecb0..5216a3c 100644
--- a/pkg/agent/runner.go
+++ b/pkg/agent/runner.go
@@ -256,7 +256,6 @@ func (r *Runner) Run(ctx context.Context, logger *zap.Logger, vmInfoUpdated util
 			c,
 			l,
 			"system",
-			func() *core.SystemMetrics { return new(core.SystemMetrics) },
 			func() bool { return true },
 			r.global.config.Metrics.System,
 			func(metrics *core.SystemMetrics, withLock func()) {
@@ -270,7 +269,6 @@ func (r *Runner) Run(ctx context.Context, logger *zap.Logger, vmInfoUpdated util
 			c,
 			l,
 			"LFC",
-			func() *core.LFCMetrics { return new(core.LFCMetrics) },
 			func() bool {
 				scalingConfig := r.global.config.Scaling.DefaultConfig.WithOverrides(getVmInfo().Config.ScalingConfig)
 				return *scalingConfig.EnableLFCMetrics // guaranteed non-nil as a required field.
@@ -369,15 +367,19 @@ func (r *Runner) spawnBackgroundWorker(ctx context.Context, logger *zap.Logger,
 	}()
 }
 
+type metricsPointer[C any, V any] interface {
+	core.FromPrometheus[C]
+	~*V
+}
+
 // getMetricsLoop repeatedly attempts to fetch metrics from the VM
 //
 // Every time metrics are successfully fetched, the value is recorded with newMetrics.
-func getMetricsLoop[M core.FromPrometheus[C], C any](
+func getMetricsLoop[M metricsPointer[C, V], C any, V any](
 	r *Runner,
 	ctx context.Context,
 	logger *zap.Logger,
 	kind string,
-	emptyMetrics func() M,
 	isActive func() bool,
 	config MetricsSourceConfig[C],
 	newMetrics func(metrics M, withLock func()),
@@ -418,14 +420,15 @@ func getMetricsLoop[M core.FromPrometheus[C], C any](
 			}
 			lastActive = true
 
-			metrics := emptyMetrics() // dependency-injected because M is usually a pointer, and otherwise the generics get too annoying.
-			err := doMetricsRequest[M, C](r, ctx, logger, timeout, metrics, config.Names)
+			var metrics V
+
+			err := doMetricsRequest[M, C](r, ctx, logger, timeout, M(&metrics), config.Names)
 			if err != nil {
 				logger.Error("Error making metrics request", zap.Error(err))
 				goto next
 			}
 
-			newMetrics(metrics, func() {
+			newMetrics(M(&metrics), func() {
 				logger.Info("Updated metrics", zap.Any("metrics", metrics))
 			})
 		}
```

The shenanigans at the bottom with repeated `M(&metrics)` _could_ get cleaner with https://github.com/samber/lo, because we could instead construct an `M` in a single line :sweat_smile: